codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parse/JsonParserGenerator.kt

```diff
@@ -0,0 +1,457 @@
+            is BlobShape -> rustTemplate("#{expect_blob_or_null}(tokens.next())?", *codegenScope)
+            is TimestampShape -> deserializeTimestamp(memberShape)
+            is CollectionShape -> deserializeCollection(target)
+            is MapShape -> deserializeMap(target)
+            is StructureShape -> deserializeStruct(target)
+            is UnionShape -> deserializeUnion(target)
+            is DocumentShape -> rustTemplate("Some(#{expect_document}(tokens)?)", *codegenScope)
+            else -> TODO(target.toString())
+        }
+        val symbol = symbolProvider.toSymbol(memberShape)
+        if (symbol.isBoxed()) {
+            rust(".map(|v| Box::new(v))")
+        }
+        if (!forceOptional && !symbol.isOptional()) {
+            if (symbol.canUseDefault()) {
+                rust(".unwrap_or_default()")
+            } else {
+                val name = escape(memberShape.memberName)
+                rustTemplate(
+                    ".ok_or_else(|| #{Error}::custom(\"value for '$name' cannot be null\"))?",
+                    *codegenScope
+                )
+            }
+        }
+    }
+
+    private fun RustWriter.deserializeString(target: StringShape) {
+        val isEnum = target.hasTrait<EnumTrait>()
+        val stringFn = if (isEnum) "expect_string_or_null" else "expect_unescaped_string_or_null"
+        rustTemplate("#{$stringFn}(tokens.next())?", *codegenScope)
+        if (isEnum) {
+            rust(".map(|s| #T::from(s.as_escaped_str()))", symbolProvider.toSymbol(target))
+        }
+    }
+
+    private fun RustWriter.deserializeNumber(target: NumberShape) {
+        val symbol = symbolProvider.toSymbol(target)
+        rustTemplate("#{expect_number_or_null}(tokens.next())?.map(|v| v.to_#{T}())", "T" to symbol, *codegenScope)
+    }
+
+    private fun RustWriter.deserializeTimestamp(member: MemberShape) {
+        val timestampFormat =
+            httpBindingResolver.timestampFormat(
+                member, HttpLocation.DOCUMENT,
+                TimestampFormatTrait.Format.EPOCH_SECONDS
+            )
+        val timestampFormatType = RuntimeType.TimestampFormat(runtimeConfig, timestampFormat)
+        rustTemplate("#{expect_timestamp_or_null}(tokens.next(), #{T})?", "T" to timestampFormatType, *codegenScope)
+    }
+
+    private fun RustWriter.deserializeCollection(shape: CollectionShape) {
+        val fnName = symbolProvider.deserializeFunctionName(shape)
+        val parser = RuntimeType.forInlineFun(fnName, "json_deser") {
+            it.rustBlockTemplate(
+                """
+                pub fn $fnName<'a, I>(tokens: &mut #{Peekable}<I>) -> Result<Option<#{Shape}>, #{Error}>
+                    where I: Iterator<Item = Result<#{Token}<'a>, #{Error}>>
+                """,
+                "Shape" to symbolProvider.toSymbol(shape),
+                *codegenScope,
+            ) {
+                startArrayOrNull {
+                    rust("let mut items = Vec::new();")
```

Comment:
at some point, we probably want to make a performance optimization to pre-allocate this vec ahead of time after pre-reading the array